### PR TITLE
Fix `==` and `!=` when comparing signed vs unsigned integers

### LINKED ISSUE #571: The result of comparing signed and unsigned variables of the same size is inconsistent
Let's take a look at the following test program:

```
a = 4294967295_u32
b = 0_i32

puts "a = #{a}"
puts "b = #{b}"

if b > a
  puts "(b > a) is true"
else
  puts "(b > a) is false"
end

if a < b
  puts "(a < b) is true"
else
  puts "(a < b) is false"
end
```

The resulting output is

```
a = 4294967295
b = 0
(b > a) is true
(a < b) is false
```

The statements "(b > a) is true" and "(a < b) is false" are obviously contradicting each other.

C/C++ compilers usually issue a warning about mixed signed/unsigned comparisons even though the language standard defines deterministic (albeit somewhat convoluted) conversion rules:
https://www.securecoding.cert.org/confluence/display/cplusplus/INT02-CPP.+Understand+integer+conversion+rules

### PR DESCRIPTION
**Pull request by asterite**
Fixes #571

### CODE REVIEW

**Comment by bcardiff** [comment]
Why closing it? I was up to send the approval.

**Comment by asterite** [comment]
@bcardiff Because it stucks at one spec, probably this broke something that depended on that equality. I'll reopen and investigate why it's failing, and try to fix that.

File: src/compiler/crystal/codegen/primitives.cr [hunk 1]
```diff
@@ -161,6 +161,26 @@ class Crystal::CodeGenVisitor
     end
   end
 
+  # The below methods (lt, lt2, gt, gte, eq, ne) perform
```
Comment (RX14) [inline src/compiler/crystal/codegen/primitives.cr]: s/lt2/lte/

**Comment by RX14** [comment]
Wow, you corrected the typo before I even had time to red the whole diff!

File: src/compiler/crystal/codegen/primitives.cr (hunk 1 above)
Comment (asterite) [reply src/compiler/crystal/codegen/primitives.cr]: Good catch. Fixed!

**Comment by asterite** [comment]
I keep adding fixes. This time, getting runtime integer values from LLVM returned `UInt64` and we sometimes compared that with `-1`, so now it fails. I fixed it by making that used method to return `Int32`, while also allowing to get the values as `Int64` and `UInt64`,  which is what LLVM provides (but in the tests we don't use such big values so it's always safe to use `Int32`).

**Comment by RX14** [comment]
CI still failing :C

**Comment by asterite** [comment]
CI passed 🎉

File: spec/std/int_spec.cr [hunk 2]
```diff
@@ -537,6 +537,18 @@ describe "Int" do
     end
   end
 
+  {% if compare_versions(Crystal::VERSION, "2.6.1") > 0 %}
```
Comment (luislavena) [inline spec/std/int_spec.cr]: @asterite is this version correct? Shouldn't be `0.26.1` instead?

Comment (asterite) [reply spec/std/int_spec.cr]: Oops! Fixed :-)

File: spec/std/int_spec.cr [hunk 3]
```diff
@@ -537,6 +537,18 @@ describe "Int" do
...
 
+  {% if compare_versions(Crystal::VERSION, "0.26.1") > 0 %}
+    it "compares equality and inequality of signed vs. unsigned integers" do
+      x = -1
+      y = x.unsafe_as(UInt32)
+
+      (x == y).should be_false
+      (y == x).should be_false
+      (x != y).should be_true
+      (y != x).should be_true
```
Comment (j8r) [inline spec/std/int_spec.cr]: Wouldn't it be better to use `should eq` and `should_not eq` expectations to have a report of variables if it fails?

Comment (asterite) [reply spec/std/int_spec.cr]: I don't know, here we are testing those operators and I wouldn't like to hide them behind a spec matcher.

Comment (RX14) [reply spec/std/int_spec.cr]: this is better as-is

**Comment by NoICE** [comment]
You guys are awesome. Thank you.

**Pull request merged**